proposed/MPID Messaging.md

```diff
@@ -0,0 +1,266 @@
+- Start Date: 22-09-2015
+- RFC PR:
+- Issue number:
+
+# Summary
+
+This RFC outlines the system components and design for general communications infrastructure and security on the SAFE Network.
+
+# Motivation
+
+## Rationale
+
+A messaging system on the SAFE Network will obviously be useful.  Over and above this, the proposed messaging system is secure and private without the possibility of snooping or tracking.
+
+A large abuse of modern day digital communications is the ability of spammers and less than ethical marketing companies to flood the Internet with levels of unsolicited email to a level of circa 90%.  This is a waste of bandwidth and a significant nuisance.
+
+## Supported Use-Cases
+
+The system will support secure, private, bi-directional communications between pairs of Clients.
+
+## Expected Outcome
+
+The provision of a secure messaging system which will eradicate unwanted and intrusive communications.
+
+# Detailed design
+
+## Overview
+
+A fundamental principle is that the cost of messaging is with the sender, primarily to deter unsolicited messages.  To achieve this, the sender will maintain messages in a network [outbox][3] until they are retrieved by the recipient.  The recipient will have a network [inbox][4] comprising a list of metadata relating to messages which are trying to be delivered to it.
+
+If the message is unwanted, the recipient simply does not retrieve the message.  The sender will thus quickly fill their own outbox with undelivered mail and be forced to clean this up themselves before being able to send further messages.
+
+This paradigm shift will mean that the obligation to unsubscribe from mailing lists, etc. is now with the owner of these lists.  If people are not picking up messages, it is because they do not want them.  So the sender has to do a better job.  It is assumed this shift of responsibilities will lead to a better-managed bandwidth solution and considerably less stress on the network and the users of the network.
+
+## Implementation Details
+
+The two relevant structs (other than the MPID itself which is a [standard Client key][0]) are the [`MpidHeader`][1] and the [`MpidMessage`][2].
+
+Broadly speaking, the `MpidHeader` contains metadata and the `MpidMessage` contains a signed header and message.  We also want to define some upper limits which will be described later.
+
+### Consts
+
+```rust
+pub const MPID_MESSAGE_TAG: u64 = 51000;
+pub const MPID_HEADER_TAG: u64 = 51001;
+pub const MAX_HEADER_METADATA_SIZE: usize = 128;  // bytes
+pub const MAX_BODY_SIZE: usize =
+    ::routing::structured_data::MAX_STRUCTURED_DATA_SIZE_IN_BYTES - 512 - MAX_HEADER_METADATA_SIZE;
+pub const MAX_INBOX_SIZE: usize = 1 << 27;  // bytes, i.e. 128 MiB
+pub const MAX_OUTBOX_SIZE: usize = 1 << 27;  // bytes, i.e. 128 MiB
+```
+
+### `MpidHeader`
+
+```rust
+pub struct MpidHeader {
+    pub sender: ::sodiumoxide::crypto::sign::PublicKey,
```

Comment:
In MpidManager's Inbox, MpidHeader will be stored as signed_header, and once client received signed_header the publick_key is required first to verify/un-patch the signed_header to MpidHeader format.
so the sender's public key shall not be defined in MpidHeader, but be stored in MpidMessage and be transfered among MpidManagers and to recipient via StructuredData's owner key (get_owner_keys() function)